src/document_graph/document.cpp

```diff
@@ -0,0 +1,270 @@
+        : contract{contract}, creator{creator}, content_groups{std::move(contentGroups)}
+    {
+        hash = hashContents();
+    }
+
+    Document::Document(eosio::name contract, eosio::name creator, ContentGroup contentGroup)
+    {
+        Document (contract, creator, rollup (contentGroup));
+    }
+
+    Document::Document(eosio::name contract, eosio::name creator, Content content)
+    {
+        Document (contract, creator, rollup (content));
+    }
+
+    Document::Document(eosio::name contract, eosio::name creator, const std::string &label, const Content::FlexValue &value)
+    {
+        Document (contract, creator, rollup (Content (label, value)));
+    }
+
+    Document::Document(eosio::name contract, const eosio::checksum256 &_hash) : contract{contract}
+    {
+        document_table d_t(contract, contract.value);
+        auto hash_index = d_t.get_index<eosio::name("idhash")>();
+        auto h_itr = hash_index.find(_hash);
+        eosio::check(h_itr != hash_index.end(), "document not found: " + readableHash(_hash));
+
+        id = h_itr->id;
+        creator = h_itr->creator;
+        created_date = h_itr->created_date;
+        certificates = h_itr->certificates;
+        content_groups = h_itr->content_groups;
+        hash = hashContents();
+
+        // this should never happen, only if hash algorithm somehow changed
+        eosio::check (hash == _hash, "fatal error: provided and indexed hash does not match newly generated hash");
+    }
+
+    void Document::emplace () 
+    {
+        require_auth(creator);
+        hash = hashContents();
+
+        document_table d_t(contract, contract.value);
+        auto hash_index = d_t.get_index<eosio::name("idhash")>();
+        auto h_itr = hash_index.find(hash);
+
+        // if this content exists already, error out and send back the hash of the existing document
+        eosio::check(h_itr == hash_index.end(), "document exists already: " + readableHash(hash));
+
+        d_t.emplace(contract, [&](auto &d) {
+            id = d_t.available_primary_key(); 
+            created_date = eosio::current_time_point();
+            d = *this;
+        });
+    }
+
+    Document Document::getOrNew (eosio::name _contract, eosio::name _creator, std::vector<ContentGroup> contentGroups) 
+    {
+        Document document {};
+        document.content_groups = contentGroups;
+        document.hash = document.hashContents();
+
+        Document::document_table d_t(_contract, _contract.value);
+        auto hash_index = d_t.get_index<eosio::name("idhash")>();
+        auto h_itr = hash_index.find(document.getHash());
+
+        // if this content exists already, return this one
+        if (h_itr != hash_index.end()) {
+            document.creator = h_itr->creator;
+            document.created_date = h_itr->created_date;
+            document.certificates = h_itr->certificates;
+            document.id = h_itr->id;
+            return document;
+        }
+
+        require_auth(_creator);
```

Comment:
I've moved it to the ```emplace``` function. We may still need to remove though because it may be called from places where we may not know who has signed the transaction, meaning whose authorization it actually has.  I need to think about this a bit more.

Reply:
I've removed ```require_auth``` from everywhere in ```document_graph```. I figure that all permissions should be handled within the contract object and related classes rather than here.